### ISSUE TITLE
Please consider the SCFA/dynamic flash attention for your implementation of scaled dot product attention

### ISSUE THREAD

**Issue by StrangeTcy** [pytorch/pytorch#103836]
### 🚀 The feature, motivation and pitch

Sparse Causal Flash Attention as implmented [here](https://github.com/epfml/dynamic-sparse-flash-attention) and described in [this paper](https://arxiv.org/abs/2306.01160) claims to 

> extend FlashAttention to accommodate a large class of attention sparsity patterns that, in particular, encompass key/query dropping and hashing-based attention. This leads to implementations with no computational complexity overhead and a multi-fold runtime speedup on top of FlashAttention. Even with relatively low degrees of sparsity, our method improves visibly upon FlashAttention as the sequence length increases. Without sacrificing perplexity, we increase the training speed of a transformer language model by 2.0× and 3.3× for sequences of respectively 8k and 16k tokens.

Adding these implementations to main Pytorch might lead to Transformers being able to use both flash attention, PEFT/LoRA and methods for handling long contexts

### Alternatives

1. Leaving everything as is and finalising the current implementation to allow handling attention masks (see issue #96099 )
2. Relying on flash_attn by HazyResearch, which works great on its own but apparently does not work with LoRA

### Additional context

I'm just not sure functions written in triton are readily convertible into C++ functions.
